types/wlr_foreign_toplevel_management_v1.c

```diff
@@ -0,0 +1,583 @@
+		struct wlr_foreign_toplevel_handle_v1 *toplevel,
+		struct wl_resource *resource) {
+	if (toplevel->title) {
+		zwlr_foreign_toplevel_handle_v1_send_title(resource,
+		    toplevel->title);
+	}
+	if (toplevel->app_id) {
+		zwlr_foreign_toplevel_handle_v1_send_app_id(resource,
+		    toplevel->app_id);
+	}
+
+	struct wlr_foreign_toplevel_handle_output *output;
+	wl_list_for_each(output, &toplevel->outputs, link) {
+		send_output_to_resource(resource, output->output, true);
+	}
+
+	struct wl_array states;
+	wl_array_init(&states);
+	bool r = fill_array_from_toplevel_state(&states, toplevel->state);
+	if (!r) {
+		wl_resource_post_no_memory(resource);
+		wl_array_release(&states);
+		return;
+	}
+
+	zwlr_foreign_toplevel_handle_v1_send_state(resource, &states);
+	wl_array_release(&states);
+
+	toplevel_update_idle_source(toplevel);
+}
+
+static void foreign_toplevel_manager_bind(struct wl_client *client, void *data,
+		uint32_t version, uint32_t id) {
+	struct wlr_foreign_toplevel_manager_v1 *manager = data;
+	struct wl_resource *resource = wl_resource_create(client,
+			&zwlr_foreign_toplevel_manager_v1_interface, version, id);
+	if (!resource) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+	wl_resource_set_implementation(resource, &foreign_toplevel_manager_impl,
+			manager, foreign_toplevel_manager_resource_destroy);
+
+	wl_list_insert(&manager->resources, wl_resource_get_link(resource));
+
+	struct wlr_foreign_toplevel_handle_v1 *toplevel, *tmp;
+	wl_list_for_each_safe(toplevel, tmp, &manager->toplevels, link) {
+		struct wl_resource *toplevel_resource =
+			create_toplevel_resource_for_resource(toplevel, resource);
+		toplevel_send_details_to_toplevel_resource(toplevel,
+		    toplevel_resource);
+	}
+}
+
+void wlr_foreign_toplevel_manager_v1_destroy(
+		struct wlr_foreign_toplevel_manager_v1 *manager) {
+	if (manager == NULL) {
+		return;
+	}
+
+	struct wlr_foreign_toplevel_handle_v1 *toplevel, *tmp_toplevel;
+	wl_list_for_each_safe(toplevel, tmp_toplevel, &manager->toplevels, link) {
+		wlr_foreign_toplevel_handle_v1_destroy(toplevel);
+	}
+
+	struct wl_resource *resource, *tmp_resource;
+	wl_resource_for_each_safe(resource, tmp_resource, &manager->resources) {
+		wl_resource_destroy(resource);
+	}
+
+	wlr_signal_emit_safe(&manager->events.destroy, manager);
+	wl_list_remove(&manager->display_destroy.link);
+
+	wl_global_destroy(manager->global);
+	free(manager);
+}
+
+static void handle_display_destroy(struct wl_listener *listener, void *data) {
+	struct wlr_foreign_toplevel_manager_v1 *manager =
+		wl_container_of(listener, manager, display_destroy);
+	wlr_foreign_toplevel_manager_v1_destroy(manager);
+}
+
+struct wlr_foreign_toplevel_manager_v1* wlr_foreign_toplevel_manager_v1_create(
+		struct wl_display *display) {
+	struct wlr_foreign_toplevel_manager_v1 *manager = calloc(1,
+			sizeof(struct wlr_foreign_toplevel_manager_v1));
+	if (!manager) {
+		return NULL;
+	}
+
+	manager->event_loop = wl_display_get_event_loop(display);
```

Comment:
No need to store this in a field, we can call it every time we need it

Reply:
(It won't return NULL)

```c
WL_EXPORT struct wl_event_loop *
wl_display_get_event_loop(struct wl_display *display)
{
	return display->loop;
}
```

Reply:
i think I saw this check somewhere else too, that's why i did it. will remove.

Reply:
yeah, found it, https://github.com/swaywm/wlroots/blob/master/types/wlr_idle.c#L227 :)

Reply:
Hmm, you still need to access the `display`, so having a field is probably fine

Reply:
didn't quite understand, is the code fine if I remove the check or not (plus if display is NULL this is crash)?

Reply:
`display` won't be NULL. Yes, please remove the check.